### ISSUE TITLE
C++ sparse_coo_tensor ignores TensorOptions argument

### ISSUE THREAD

**Issue by meganset** [pytorch/pytorch#55453]
## 🐛 Bug

C++ sparse_coo_tensor ignores TensorOptions argument when supplied with indices & values

## To Reproduce

sparse_coo_tensor has three signatures:
```
TORCH_API Tensor sparse_coo_tensor(IntArrayRef size, TensorOptions options);
TORCH_API Tensor sparse_coo_tensor(const Tensor & indices, const Tensor & values, TensorOptions options={});
TORCH_API Tensor sparse_coo_tensor(const Tensor & indices, const Tensor & values, IntArrayRef size, TensorOptions options={});
```
Any tensor options given in the size-only call are used, but the others seem to ignore any options other than to check if the layout contradicts the sparse intent.
e.g. in the size only case, the empty tensor is created on CUDA w'datatype=double:

``` 
 auto o=torch::dtype(torch::kDouble).device(torch::kCUDA);
 auto t1=torch::sparse_coo_tensor({2,3}, o);
 std::cerr << t1 << "\n";

[ SparseCUDADoubleType{}
indices:
[ CUDALongType{2,0} ]
values:
[ CUDADoubleType{0} ]
size:
[2, 3]
]
```

When indices and values supplied, the TensorOptions are ignored:
```
 auto i=torch::tensor({{1},{1}});
 auto v=torch::tensor({1});
 auto t2=torch::sparse_coo_tensor(i,v,o);
 std::cerr << t2 << "\n";

[ SparseCPULongType{}
indices:
 1
 1
[ CPULongType{2,1} ]
values:
 1
[ CPULongType{1} ]
size:
[2, 2]
]
```

Thanks for looking into this.

**Comment by mruberry** [comment]
Thank you for reporting this issue, @meganset. cc @pearu, one of our sparse experts who's also be looking at tensor creation operations recently.

**Comment by pearu** [comment]
In the case of tensor creation functions such as `torch.as_tensor` or `torch.tensor`, the user-supplied tensor options such as `dtype` and `device` are applied to the input tensor.

I think applying the same convention to constructing sparse tensors makes sense as well. That is, the following transformation should be effective when a user specifies tensor options:
```
torch.sparse_coo_tensor(indices, values, options=options) -> 
torch.sparse_coo_tensor(indices.to(device=options.device),
                        values.to(device=options.device,
                                  dtype=options.dtype,
                                  requires_grad=options.requires_grad))
```
Notice that options like `dtype` and `requires_grad` are applied only to `values` because `indices.dtype` must be `int64` and `indices` are not differentiable.

The tensor option `layout` can be either unspecified or `torch.strided`, otherwise an exception should be raised because currently the COO indices/values tensors must have strided layout.

Does this specification make sense?

**Comment by meganset** [comment]
Applying any specified device to the indices
and the device, dtype & any gradient requirement to the values sounds good to me.

The current c++ code seems reasonable as far as layout:
```
  TORCH_CHECK(!options.has_layout() || options.layout() == kSparse, "expected sparse layout, but got layout ", options.layout());
```
Requiring the layout specification to be _strided_ (with the layout applying to the input tensor of values)
seems confusing,
especially in the call where only size & options are expected:
```
TORCH_API Tensor sparse_coo_tensor(IntArrayRef size, TensorOptions options);
```

Thanks for looking at this.

**Comment by pearu** [comment]
Thanks, @meganset for pointing out the current `layout` treatment. I have updated the specs accordingly.
